### ISSUE TITLE
N-Rage Modifiers Macro not working

### ISSUE THREAD

**Issue by Dragorn421** [project64/project64#1711]
Hi, I'm trying to use an xbox controller in project64, that works fine, but also have some buttons also usable with keyboard.
So for the keyboard part **I tried "A key presses A button" with the modifiers->macros feature (screenshots below) but it just won't work.**

I tried these two versions, with toggle on and off, with and without background inputs in nightly, and never got it to work. Screenshots taken after reopening controls window so it's **not a controls-not-saved issue**:
2.3.2.202 (the one I had installed)
![image](https://421.es/doyu/1ddjhs)

**v2.4.0-1127-g185c658 (latest nightly as time of writing)**
![image](https://421.es/doyu/1ddm5c)

ROM used: Ocarina of Time MQ debug (ZELOOTMA.Z64) (it has an in-game input display so I also know N-Rage was actually not pressing any button, so it's **not an issue of wrong button pressed**)
My keyboard is Azerty.

Feel free to ask me more details if needed.

**Comment by IronYoshi64** [comment]
Did you try clicking the Rapid Fire button, and/or tick the Slow Rapid Fire option, then click Apply Changes, then Save?

**Comment by Dragorn421** [comment]
Yes, on both version, doesn't help :(

**Comment by Dragorn421** [comment]
I should have tried this way sooner but disabling XInput makes the modifiers macro work. So I guess there's a bug here: enabling XInput disables modifiers macros

**Comment by IronYoshi64** [comment]
Wow OK. That was eventually going to be my next suggestion. Well yeah glad you figured it out. That is definitely a bug then. Maybe you or someone else should rename this issue to something like "XInput Disables Modifier Macros" to be more clear.

**Comment by legend800** [comment]
I can confirm this issue is still active. 

Sadly, there's no way to invert C-buttons when using Xinput, so we're forced to use macros for that use case and similar but as this bug mentions, that's blocked currently.

**Comment by xmat123x** [comment]
This is still an issue, wish there was some kind of workaround.

**Comment by Squall-Leonhart** [comment]
use a dinput controller.

**Comment by legend800** [comment]
That's not a solution and a worthless comment.

**Comment by Squall-Leonhart** [comment]
incorrect, its the only solution and a perfectly valid one at that, and since i don't see you committing or providing code to an alternative, shut your mouth.

**Comment by Squall-Leonhart** [comment]
> Wow OK. That was eventually going to be my next suggestion. Well yeah glad you figured it out. That is definitely a bug then. Maybe you or someone else should rename this issue to something like "XInput Disables Modifier Macros" to be more clear.

There is no bug here, the modifiers are only implemented via dinput for that dinput code.
Ticking Xinput disables this intentionally on the Xinput enabled controller to prevent double inputs and double button presses, at the point of time nrage was made Dinput was still the common controller available on the market.

A physical Dinput controller or virtualizing one with Xinputplus are the only way and nobody cares enough about nrage anymore to implement macro's with Xinput.

**Comment by legend800** [comment]
Wow, your solution is obvious to anyone. The bug is about xinput so obviously dinput works fine. That's not the point. Do you go in every bug and saw just don't do that thing that triggers the issue? 

The point is there a software defect. It either should work or the functionally should not be available\grayed out. You're also telling me it's impossible to make macros work with xinput? It could,. It's a current limitation of them plugin if we want to argue about labels.

**Comment by Squall-Leonhart** [comment]
"Doctor, it hurts when i punch myself in the face"
"So don't punch yourself in the face'

Its not a defect, its a feature that was never added.

**Comment by S-T-A-R-S** [comment]
> "Doctor, it hurts when i punch myself in the face"
> "So don't punch yourself in the face'
> 
> Its not a defect, its a feature that was never added.

Add it then, you rude jackass. People want it. why are you so bitter in every thread I see you in? jesus.

**Comment by Squall-Leonhart** [comment]
mmmmmm

No. do it yourself.

**Comment by Bobsson** [comment]
> The point is there a software defect. It either should work or the functionally should not be available\grayed out. You're also telling me it's impossible to make macros work with xinput? It could,. It's a current limitation of them plugin if we want to argue about labels.

This.  It's irrelevant if there's a viable "workaround" or not - the functionality should either work or not be available so that users know it won't work and can consider other options.  I can understand the actual functionality being more complicated to implement than anyone wants to deal with, but disabling this screen when the XInput checkbox is checked should be a pretty easy fix.

**Comment by Squall-Leonhart** [comment]
all this and you people still haven't figured out this isn't the nrage repo.

**Comment by Dragorn421** [comment]
Isn't this the nrage repo https://github.com/project64/project64/tree/develop/Source/nragev20

**Comment by Squall-Leonhart** [comment]
no

**Comment by Dragorn421** [comment]
how many people will need to tell you you're a prick before you decide to act nicer

where is the nrage repo then?

**Comment by flagrama** [comment]
https://github.com/project64/project64/tree/develop/Source/nragev20 is close enough.

All of the plugins distributed with the emulator except GLideN64 are maintained in this repo since development of most of them has slowed or stopped otherwise.

**Comment by Dragorn421** [comment]
Yeah that's what I figured, the sourceforge page which seems to be the original page hasn't been updated in years.
So this is relevant to project64/project64 afaict.

I should check if this is even still an issue, as it's 2 years+ old. In that time I may even have learned enough to fix it myself who knows

Thanks for the helpful answer flagrama.

**Comment by Pizzanomicon** [comment]
Using Project 64 v2.3.2 this appears to still be an issue.  As a newcomer to this thread, I can say this request seems valid-  insofar largely as it is one I share.  Using a Xinput controller is an increasingly common way to interface with Windows "gaming" machines, and I think the devs behind Project64 saw that coming years ago. 

I also think as an accessibility feature, more people are turning to emulators to easily provide support for workarounds to button-mashing, which becomes difficult and painful for some folk.  

It does appear the N-Rage plugin (v2) does have a makefile now, so I'm going to see about compiling this on my own...

**Comment by Squall-Leonhart** [comment]
this request is not valid and will not receive any further attention.

**Comment by DerekTurtleRoe** [comment]
Pinging everybody in the thread who was seemingly having issues with the plugin: @Pizzanomicon @Dragorn421 @Bobsson @S-T-A-R-S @legend800 @xmat123x 

Firstly I would like to say that if you are experiencing issues, the only versions I can recommend you try are the latest stable (3.0.1) and the latest development build (4.x). Older versions we aren't going to be maintaining and support for older operating systems and older hardware gets more and more difficult every day, at least if that is why you are using older versions. If you have a modern enough system, please upgrade your version and try again and see if is still an issue! 😄 

Secondly, N-Rage is not really maintained anymore, and so I only recommend it for people who are using an XInput controller, or want things like Transfer Pak support or other more obscure features. Project64-input was a new plugin made to be more simple and contain a majority of the features but with far less complexity, and so it is good, but it only works with DirectInput controllers. I have spoken to Zilmar about this and he agreed that at some point we would like to add support for XInput to that if possible. When that will be, we don't know. But we are aware that both plugins have pros and cons.

Typically if I see somebody having issues, there is another fairly new plugin available [here](https://github.com/wermipls/octomino-sdl-input) and it is what I'm using myself and recommending for the most part. The only exception being for Transfer Pak support and development and testing, but I assume most of you don't use or need plugins for that. 😅 

Lastly, I do apologize for the way Squall has been acting in the thread and others. But know that I do read these threads and I try to test, retest, and verify the things you guys post here as issues, and I do talk with others about it as well. I don't always have the time I'd like to answer messages and emails and post in Discord, but I do my best. 😅 

Just for reference, it is mostly me, Zilmar, and Squall who do testing and Zilmar does most of the development. I do as much as I can, but me and Zilmar have other things that bring us income and we focus on those typically. Zilmar also has a family to take care of and I have had some personal things going on lately and I just haven't had a lot of free time. So I do apologize that it might seem like we aren't doing anything, but as I said I am doing things in the background typically to help out. Sometimes issues just go above my head and I am not sure how to approach them. Hopefully you all can understand that, at least.

Anyway, sorry for the wall of text and pings, but just wanted to put this out there.

Thanks,

Turtle

**Comment by legend800** [comment]
Thanks Turtle for all the details. 

I tried that plugin and while I had full control over stick mapping, seems like it doesn't support rumble. 

For me, **_Xinput Plus_** is still the only way to have rumble and be able to invert axis of the sticks.
